Tp2000 1371 implement new sub quotas journey

CPrich905 opened the pull request:
# TP-2000-1371 Implement new sub quotas journey

## Why
Currently to create an association between quotas, TMs must follow the following process:
1. create both main and sub quota order numbers, 
8. create two sets of definitions, one on each order number
9. raise a ticket for the Data team to create the association between the definitions.

This has a number of issues, namely it is prone to introducing errors by manually inputting the same data for each set of definitions, requires the Data team to directly manipulate the database, adding a further risk for introducing errors.

## What
The Tariff Manager is now guided through a form wizard that:
1. selects the main and sub quota order numbers (these must be created before starting the journey)

<img width="1246" alt="image" src="https://github.com/user-attachments/assets/f60316d4-8fda-4b9a-804a-6ad25273e824">

2. selects the definitions to be duplicated from main to sub quota order number
<img width="1209" alt="image" src="https://github.com/user-attachments/assets/afbebd0b-2e5c-48a8-b326-7b2f28ba87d7">

3. for each definition selected: provides the relationship type and coefficient values for the association between main and sub definition, makes any updates required to the sub-definition. This information is stored on a new QuotaDefinitionDuplicator table.
<img width="1216" alt="image" src="https://github.com/user-attachments/assets/1c3c92b0-9bee-4896-b093-fa34e275db57">

<img width="1230" alt="image" src="https://github.com/user-attachments/assets/f138f851-db84-491e-9d2f-bb387ea07ef5">

4. on submission, the quota definitions are duplicated and for each definition pair, an entry is made into the QuotaAssociation table

## Checklist
- Requires migrations? Yes
- Requires dependency updates? No

Links to relevant material
See: [Quota Association in the Tariff Data Manual](https://uktrade.github.io/tariff-data-manual/documentation/data-structures/quota-associations.html)

eadpearce commented on quotas/forms.py:
```diff
@@ -1001,3 +1006,300 @@ def create_blocking_period(self, workbasket):
+    pass
+
+
+class QuotaOrderNumersSelectForm(forms.Form):
```
`QuotaOrderNum**b**ersSelectForm` typo

TomMacca reviewed (approved):
Looks good, me and Steve want to try this out so have deployed on dev. Black tests were failing but it was some nonsense now fixed. I corrected the little typo Edie pointed out. Sorry to barge in @CPrich905 its just your feature looks great and might really help us out with steel stuff 🙌

Edit: 

Actually we found that it was flagging rule violations when we were trying to associate two quotas that were both 'NM's

SteveCN7 commented:
On Friday, @TomMacca and I used this PR in it's current state to attempt to resolve a test file setup for Steel Safeguards. 
Creating the definition periods on the association as well the association itself went flawless when associating the first quota to the main. 
The second quota association however, appeared to fail validation on screen, as it failed on association to NM quotas, referencing that one of them wasn't an NM quota.

CPrich905 commented:
> Looks good, me and Steve want to try this out so have deployed on dev. Black tests were failing but it was some nonsense now fixed. I corrected the little typo Edie pointed out. Sorry to barge in @CPrich905 its just your feature looks great and might really help us out with steel stuff 🙌
> 
> Edit:
> 
> Actually we found in Dev that the feature is flagging rule violations when we were trying to associate two quotas that were both 'NM's

@TomMacca this issue is now resolved - I've also expanded the business rules checks (I'd missed a requirement for QA5) and added further tests for the business rules

mattjamc commented on quotas/forms.py:
```diff
@@ -1005,3 +1017,313 @@ def create_blocking_period(self, workbasket):
+        help_text=COEFFICIENT_HELP_TEXT,
+        error_messages={
+            "invalid": "Coefficient must be a number",
+            "required": "Enter the volume",
```
Should this be "Enter the coefficient"

mattjamc commented on quotas/forms.py:
```diff
@@ -1005,3 +1017,313 @@ def create_blocking_period(self, workbasket):
+    )
+
+    initial_volume = forms.DecimalField(
+        label="Volume",
```
Should this label be "Initial volume"

mattjamc commented on quotas/forms.py:
```diff
@@ -1005,3 +1017,313 @@ def create_blocking_period(self, workbasket):
+    def init_fields(self):
+        self.fields["measurement_unit"].queryset = self.fields[
+            "measurement_unit"
+        ].queryset.order_by("code")
```
Would it make more sense to add the .order_by("code") to where the original queryset is defined on line 1181 rather than replace the queryset here

mattjamc commented on quotas/jinja2/quota-definitions/sub-quota-definitions-select-definition-period.jinja:
```diff
@@ -0,0 +1,63 @@
+    {% set main_quota_order_number = view.get_cleaned_data_for_step('quota_order_numbers')['main_quota_order_number']%}
+    {% set sub_quota_order_number = view.get_cleaned_data_for_step('quota_order_numbers')['sub_quota_order_number']%}
+    <h2>Select definition periods from main quota ID {{main_quota_order_number}}</h2>
+    <p>These definition periods are required for sub-quota ID {{sub_quota_order_number}} and must be edited on the next pages</p>
```
Could this <p> be clearer unless I am misunderstanding. I'm not sure what it means by 'These definition periods are required for sub-quota ID XXX'.

mattjamc replied:
Ah right that's what I ended up thinking it means! I think that makes more sense but happy to leave it

mattjamc commented on quotas/jinja2/quota-definitions/sub-quota-definitions-selected.jinja:
```diff
@@ -0,0 +1,68 @@
+{% extends "quota-definitions/sub-quota-duplicate-definitions-step.jinja"%}
+{% from "components/create-status-tag.jinja" import create_status_tag %}
```
Is create_status_tag needed if we are using the status_tag_generator instead

mattjamc commented on quotas/jinja2/quota-definitions/sub-quota-definitions-selected.jinja:
```diff
@@ -0,0 +1,68 @@
+      ]) or "" }}
+      
+    {% endfor %}
+    <h2>You must enter a co-efficient value and specify the relationship type for each definition period.</h2>
```
Maybe this is better as a paragraph as it's not really a heading but more help text

mattjamc commented on quotas/jinja2/quota-definitions/sub-quota-definitions-selected.jinja:
```diff
@@ -0,0 +1,68 @@
+      ],
+      "rows": table_rows,
+    }) }}
+    <p>Selecting 'Submit' will create the new definitions and create a relationship in QuotaAssociations. Further edits to the definitions can be made via the Quota view</p>
```
This makes sense to us but I'm not sure the users will understand 'create a relationship in QuotaAssociations'. Maybe better saying that it will simply create an association (or however they might phrase it).

mattjamc commented on quotas/jinja2/quota-definitions/sub-quota-definitions-selected.jinja:
```diff
@@ -0,0 +1,68 @@
+    {{ govukButton({
+    "text": "Submit",
+  }) }}
+  {{ govukButton({
```
Maybe add these two inside a <div class="govuk-button-group">

CPrich905 commented on quotas/jinja2/quota-definitions/sub-quota-definitions-select-definition-period.jinja:
```diff
@@ -0,0 +1,63 @@
+    {%- endset %}
+
+    {% set start_date %}
+      {{ create_sortable_anchor(request, "valid_between", "Start date", base_url) }}
```
I should have checked this more thoroughly. Because we're not getting these through a url request, we can't use this method for sorting the definition periods. I've removed this for now but will have to investigate further and add it in a later PR.

mattjamc commented on quotas/jinja2/quota-definitions/sub-quota-definitions.jinja:
```diff
@@ -0,0 +1,17 @@
+{% set page_title = "Select definition periods to duplicate (sub-quota-definitions.jinja)" %}
```
Is this file being used at all?

CPrich905 replied:
Nope, removed

mattjamc commented on quotas/jinja2/quota-definitions/sub-quota-duplicate-definitions-step.jinja:
```diff
@@ -0,0 +1,24 @@
+{% extends "layouts/form.jinja" %}
```
Just wondering when this file is used - I see it's the default template in _get_template_names_ if a template for a step is not found but when would that happen?

mattjamc commented on quotas/serializers.py:
```diff
@@ -262,3 +283,42 @@ class Meta:
+    end_date = deserialize_date(definition.definition_data["end_date"])
+    initial_volume = Decimal(definition.definition_data["initial_volume"])
+    vol = Decimal(definition.definition_data["volume"])
+    measurement_unit = measures.models.tracked_models.MeasurementUnit.objects.get(
```
Should we just import MeasurementUnit if that's the only thing we are using rather than all of measures?

CPrich905 replied:
Yes, yes we should

mattjamc commented on quotas/serializers.py:
```diff
@@ -262,3 +283,42 @@ class Meta:
+    measurement_unit = measures.models.tracked_models.MeasurementUnit.objects.get(
+        code=definition.definition_data["measurement_unit_code"]
+    )
+    sub_order_number = self.get_cleaned_data_for_step(self.QUOTA_ORDER_NUMBERS)[
```
Just a question - why is the rest of the data being serialized, stored in the model then being pulled out but for the sub_order_number it's calling get_cleaned_data_for_step and getting it that way?

CPrich905 replied:
I can't recall if there was a particular reason for not including this on the model at the time... but no particular reason I don't think!

CPrich905 commented on quotas/jinja2/quota-definitions/sub-quota-definitions-updates.jinja:
```diff
@@ -0,0 +1,30 @@
+{% extends "layouts/form.jinja" %}
+{% from "components/table/macro.njk" import govukTable %}
+{% from "components/details/macro.njk" import govukDetails %}
```
Nope, removed!

CPrich905 commented on quotas/forms.py:
```diff
@@ -1005,3 +1017,300 @@ def create_blocking_period(self, workbasket):
+        fields["end_date"].initial = deserialize_date(duplicate_data["end_date"])
+
+    def init_fields(self):
+        self.fields["measurement_unit"].queryset = self.fields[
```
You are correct, removed!

mattjamc commented on quotas/views.py:
```diff
@@ -732,6 +736,264 @@ class QuotaDefinitionConfirmDelete(
+        context["quota_order_number"] = self.kwargs["sid"]
+        return context
+
+    def get_main_definition(self):
```
Could do this as a @property called main_definition if we are just returning the quotadefinition object (or can leave it)

CPrich905 replied:
I actually started off with this using the decorator but changed it during a refactor. This works well enough, so will leave as is unless someone presents an argument in favour of changing it back :)

mattjamc commented on quotas/tests/test_views.py:
```diff
@@ -1688,3 +1691,236 @@ def test_quota_blocking_confirm_create_view(valid_user_client):
+    assert response.status_code == 200
+
+
+def test_definition_duplicator_creates_definition_and_associaion(
```
association typo

mattjamc commented on quotas/tests/test_views.py:
```diff
@@ -1688,3 +1691,236 @@ def test_quota_blocking_confirm_create_view(valid_user_client):
+    """
+    Pass data to the Duplicator Wizard and verify that the created definition
+    contains the expected data.
+    This must include: quota_order_numbers,
```
Is this docstring finished?

mattjamc commented on quotas/tests/test_views.py:
```diff
@@ -1688,3 +1691,236 @@ def test_quota_blocking_confirm_create_view(valid_user_client):
+    )[0]
+
+    association_table_before = models.QuotaAssociation.objects.all()
+    association_table_before = models.QuotaAssociation.objects.all()
```
Line repeated

mattjamc reviewed (approved):
Looking good!

CPrich905 commented on quotas/models.py:
```diff
@@ -500,3 +500,33 @@ class QuotaEvent(TrackedModel):
     business_rules = (UpdateValidity,)
+
+
+class QuotaDefinitionDuplicator(models.Model):
```
It could be, however when I was talking this through at the start of the project with @eadpearce we felt that using a model to store the data would be a better option, primarily because this could be a long/laborious process and were the journey interrupted and browser closed, the user's progress would be lost.

mattjamc commented on quotas/jinja2/quota-definitions/sub-quota-definitions-selected.jinja:
```diff
@@ -0,0 +1,69 @@
+    
+    {% for definition in data %}
+      {% set edit_link -%}
+        <a class="govuk-link govuk-!-font-weight-bold" href="{{ url("sub_quota_definitions-ui-updates", kwargs={"sid": definition.main_definition.trackedmodel_ptr_id}) }}">Edit</a>
```
As this kwarg is sid should we not be passing the quota definition sid in rather than the trackedmodel_ptr_id?

CPrich905 replied:
Corrected this to <pk>

paulpepper-trade commented on quotas/views.py:
```diff
@@ -732,6 +736,242 @@ class QuotaDefinitionConfirmDelete(
+):
+    """UI endpoint for any updates to duplicated definitions"""
+
+    storage_name = "quotas.wizard.QuotaDefinitionDuplicatorSessionStorage"
```
Is `storage_name` used anywhere in this Form class?

CPrich905 replied:
Removed

paulpepper-trade commented on quotas/views.py:
```diff
@@ -732,6 +736,241 @@ class QuotaDefinitionConfirmDelete(
+        """Parses and converts a date string from that used for storing data
+        to the one used in the TAP UI."""
+        if date_str:
+            date_object = datetime.datetime.strptime(date_str, "%Y-%m-%d").date()
```
We should use [TAP's standard date formats defined in the project's settings](https://github.com/uktrade/tamato/blob/95b34157b927753e27479c2e894e1edce71e2d34/settings/common.py#L412-L417).

CPrich905 replied:
As discussed offline

CPrich905 commented on quotas/views.py:
```diff
@@ -732,6 +736,241 @@ class QuotaDefinitionConfirmDelete(
+            sub_quota=cleaned_data["sub_quota_order_number"],
+            definition_view_url=(f"{sub_quota_view_url}?{sub_quota_view_query_string}"),
+        )
+        return render(
```
Corrected to redirect

paulpepper-trade commented on quotas/views.py:
```diff
@@ -732,6 +737,253 @@ class QuotaDefinitionConfirmDelete(
+            ),
+        }
+
+        return redirect(reverse("sub_quota_definitions-ui-success"))
```
It doesn't do any harm, but `reverse()` isn't actually required as `redirect()` [will accept a view name](https://docs.djangoproject.com/en/4.2/topics/http/shortcuts/#redirect). We seem to have a combination of styles through the codebase!

CPrich905 replied:
Good to know! I'd followed an example which used both, but if I've understood the docs better on my re-read this morning, `redirect()` uses the `reverse()` method anyway - should we be encouraging the use of `redirect()` on its own as standard?
